### ISSUE TITLE
"idx" problem in MPNN

### ISSUE THREAD

**Issue by Frank-LIU-520** [rusty1s/pytorch_geometric#2003]
## ❓ Questions & Help
In the exapmle code of qm9_nn_conv.py, why should we use "idx" to index our edge_attr.
Can we remove `**idx = data.edge_index[0] * data.num_nodes + data.edge_index[1]**` and `edge_attr[idx] = data.edge_attr`? Or can we just use `edge_attr = data.edge_attr`  since we can directly call edge_attr with a number, like edge_attr[0] or any integer number to call the edge attr.

Here is the codes used for data transformation below.
`class Complete(object):
    def __call__(self, data):
        device = data.edge_index.device

        row = torch.arange(data.num_nodes, dtype=torch.long, device=device)
        col = torch.arange(data.num_nodes, dtype=torch.long, device=device)

        row = row.view(-1, 1).repeat(1, data.num_nodes).view(-1)
        col = col.repeat(data.num_nodes)
        edge_index = torch.stack([row, col], dim=0)

        edge_attr = None
        if data.edge_attr is not None:
            **idx = data.edge_index[0] * data.num_nodes + data.edge_index[1]**
            size = list(data.edge_attr.size())
            size[0] = data.num_nodes * data.num_nodes
            edge_attr = data.edge_attr.new_zeros(size)
            **edge_attr[idx] = data.edge_attr**

        edge_index, edge_attr = remove_self_loops(edge_index, edge_attr)
        data.edge_attr = edge_attr
        data.edge_index = edge_index

        return data
`

**Comment by rusty1s** [comment]
`idx` contains the indices of where elements of `data.edge_attr` should be placed in a flattened *dense* adjacency matrix of shape `[num_nodes * num_nodes]`. This is needed because here, we operate on a full graph (each node connected to every other node), while we want to maintain the correct edge features for the initial nodes.

**Comment by Frank-LIU-520** [comment]
> 
> 
> `idx` contains the indices of where elements of `data.edge_attr` should be placed in a flattened _dense_ adjacency matrix of shape `[num_nodes * num_nodes]`. This is needed because here, we operate on a full graph (each node connected to every other node), while we want to maintain the correct edge features for the initial nodes.


Thanks for your reply. I am still little confused about the meaning of using `idx` in `edge_attr[idx]`.  
Suppose we have one molecular graph with 39 nodes, each node has 1*186 feature. Let us print the `idx` 、`data.edge_index,` `data.edge_attr`  and we can found the the shape of idx is 2601 ,  [2, 2601] for data.edge_index and [2601, 40] for data.edge_attr.
So my question goes that why should we need `edge_attr[idx]` while we already have `data.edge_index` for fully-connected adjacent matrix? What functions can we trace the whole computation process of the use of `idx`?


Codes are as follows:

```
print("Num of nodes", data.x.shape)   #output : Num of nodes torch.Size([39, 186])
if data.edge_attr is not None:
        idx = data.edge_index[0] * data.num_nodes + data.edge_index[1]
        print("idx", idx.shape, idx)              # output : idx torch.Size([2601]) tensor([   0,    1,    2,  ..., 2598, 2599, 2600])
        size = list(data.edge_attr.size())
        size[0] = data.num_nodes * data.num_nodes
        edge_attr = data.edge_attr.new_zeros(size)
        print("edge_index",edge_index.shape)   #outpu: edge_index  torch.Size([2, 2601])
        print("edge_attr",edge_attr.shape)   #output:  edge_attr  torch.Size([2601, 40])
        edge_attr[idx] = data.edge_attr  !!! notice where should be changed here
        print("edge_attr[idx]", edge_attr[idx].shape) #output : edge_attr[idx]  torch.Size([2601, 40])

edge_index, edge_attr = remove_self_loops(edge_index, edge_attr)
data.edge_attr = edge_attr
data.edge_index = edge_index
return data

```
where should be wrong if I wrote these code in the following way?

Codes are as follows:
```
print("Num of nodes", data.x.shape)   #output : Num of nodes torch.Size([39, 186])
if data.edge_attr is not None:
        size = list(data.edge_attr.size())
        size[0] = data.num_nodes * data.num_nodes
        edge_attr = data.edge_attr.new_zeros(size)
        print("edge_index",edge_index.shape)   #outpu: edge_index  torch.Size([2, 2601])
        print("edge_attr",edge_attr.shape)   #output:  edge_attr  torch.Size([2601, 40])
        **edge_attr= data.edge_attr**   !!! notice the change here

edge_index, edge_attr = remove_self_loops(edge_index, edge_attr)
data.edge_attr = edge_attr
data.edge_index = edge_index

return data
```

**Comment by rusty1s** [comment]
The above computations converts a sparsely-connected graph into a fully-connected graph. Say, for example, you have the following graph:
```python
edge_index = [[0, 1], [1, 2]]
edge_attr = [1, 2]
```
Now, to convert it into a fully connected, there are some edges missing, namely `(0, 0), (0, 2), (1, 0), (1, 1), (2, 0), (2, 1), (2, 2)`
resulting in the following `edge_index`:
```python
edge_index = [[0, 0, 0, 1, 1, 1, 2, 2, 2], [0, 1, 2, 0, 1, 2, 0, 1, 2]]
```
These changes in the `edge_index` needs to also be reflected in `edge_attr`. We want to maintain the edge attributes of the original graph, and want to add new zero attributes for the newly introduced edges:
```python
edge_attr = [0, 1, 0, 0, 0, 2, 0, 0, 0]
```
This is exactly where `idx` comes into play, since it holds the position/indices of the original edges in the fully-connected graph. Namely, `idx = [1, 5]`.

**Comment by Frank-LIU-520** [comment]
> 
> 
> The above computations converts a sparsely-connected graph into a fully-connected graph. Say, for example, you have the following graph:
> 
> ```python
> edge_index = [[0, 1], [1, 2]]
> edge_attr = [1, 2]
> ```
> 
> Now, to convert it into a fully connected graph, there are some edges missing, namely `(0, 0), (0, 2), (1, 0), (1, 1), (2, 0), (2, 1), (2, 2)`, resulting in the following `edge_index`:
> 
> ```python
> edge_index = [[0, 0, 0, 1, 1, 1, 2, 2, 2], [0, 1, 2, 0, 1, 2, 0, 1, 2]]
> ```
> 
> These changes in the `edge_index` needs to also be reflected in `edge_attr`. We want to maintain the edge attributes of the original graph, and want to add new zero attributes for the newly introduced edges:
> 
> ```python
> edge_attr = [0, 1, 0, 0, 0, 2, 0, 0, 0]
> ```
> 
> This is exactly where `idx` comes into play, since it holds the position/indices of the original edges in the fully-connected graph. Namely, `idx = [1, 5]`.

Got it, thanks you so much Matthias.  Now I am so happy that I finally know the meaning of idx with your help.
